src/app/shared/input/abstract-input.ts

```diff
@@ -0,0 +1,66 @@
+import { Directive, ElementRef, Input, OnInit, ViewChild } from '@angular/core';
+import { ControlValueAccessor } from '@angular/forms';
+
+@Directive()
+export class AbstractInputComponent<T> implements OnInit, ControlValueAccessor {
+  @ViewChild('inputField')
+  inputField: ElementRef;
+
+  constructor() {}
+
+  onChange = (newValue: T) => {};
+
+  onTouched = () => {};
+
+  writeValue(newValue: any): void {
+    this.value = newValue;
+  }
+
+  registerOnChange(fn: any): void {
+    this.onChange = fn;
+  }
+
+  registerOnTouched(fn: any): void {
+    this.onTouched = fn;
+  }
+
+  setDisabledState?(isDisabled: boolean): void {
+    this.disabled = isDisabled;
+  }
+
+  focus() {
+    if (this.inputField && this.inputField.nativeElement) {
+      this.inputField.nativeElement.focus();
+    }
+  }
+
+  @Input()
```

Comment:
Consider placing @[Input](https://angular.io/api/core/Input)() or @[Output](https://angular.io/api/core/Output)() on the same line as the property it decorate. And declare them in the top, before any function declarations

Reply:
Done